synapse/lib/layer.py

```diff
@@ -4093,6 +4094,86 @@ async def iterNodeDataKeys(self, buid):
             prop = self.getAbrvProp(abrv)
             yield prop[0]
 
+    async def iterLayerAddPerms(self):
+
+        # nodes & props
+        for byts, abrv in self.propabrv.slab.scanByFull(db=self.propabrv.name2abrv):
+            form, prop = s_msgpack.un(byts)
+            if form is None:
+                continue
+
+            if await self.layrslab.countByPref(abrv, db=self.byform):
+                yield ('node', 'add', form)
+
+            if prop and await self.getPropCount(form, prop):
+                yield ('node', 'prop', 'set', f'{form}:{prop}')
+
+        # tags
+        seen = await s_spooled.Dict.anit(cell=self.core)
+
+        for tag in self.tagabrv.names():
+            if not await self.getTagCount(tag):
+                continue
+
+            parts = tag.split('.')
+            for idx in range(1, len(parts) + 1):
+                key = tuple(parts[:idx])
+                await seen.set(key, seen.get(key, 0) + 1)
+
+        for key, count in seen.items():
+            if count == 1:
+                yield ('node', 'tag', 'add', *key)
+
+        # tagprops
+        for byts, abrv in self.tagpropabrv.slab.scanByFull(db=self.tagpropabrv.name2abrv):
+            info = s_msgpack.un(byts)
+            if None in info or len(info) != 3:
+                continue
+
+            form, tag, prop = info
+            if not await self.getTagPropCount(form, tag, prop):
+                continue
+
+            yield ('node', 'tag', 'add', *tag.split('.'))
+
+        # nodedata
+        for abrv, _ in self.dataslab.scanByFull(db=self.dataname):
+            if not await self.dataslab.countByPref(abrv, db=self.dataname):
```

Comment:
Is it possible to have nodedata in the slab that isn't referenced? Should I remove this check?